oshw/erika/nicdrv.c

```diff
@@ -0,0 +1,562 @@
+      			port->stack.rxbufstat   = &(port->rxbufstat);
+      			port->stack.rxsa        = &(port->rxsa);
+      			ecx_clear_rxbufstat(&(port->rxbufstat[0]));
+
+   			/* setup ethernet headers in tx buffers so we don't have to repeat it */
+   			for (i = 0; i < EC_MAXBUF; i++)
+   			{
+      				ec_setupheader(&(port->txbuf[i]));
+      				port->rxbufstat[i] = EC_BUF_EMPTY;
+   			}
+   			ec_setupheader(&(port->txbuf2));
+
+			break; // device found
+		}
+	}
+
+	return (dev != NULL);
+}
+
+/** Close sockets used
+ * @param[in] port        = port context struct
+ * @return 0
+ */
+inline int ecx_closenic(ecx_portt *port)
+{
+	// Nothing to do
+   	return 0;
+}
+
+/** Fill buffer with ethernet header structure.
+ * Destination MAC is always broadcast.
+ * Ethertype is always ETH_P_ECAT.
+ * @param[out] p = buffer
+ */
+void ec_setupheader(void *p)
+{
+   	ec_etherheadert *bp;
+   	bp = p;
+   	bp->da0 = oshw_htons(0xffff);
+   	bp->da1 = oshw_htons(0xffff);
+   	bp->da2 = oshw_htons(0xffff);
+   	bp->sa0 = oshw_htons(priMAC[0]);
+   	bp->sa1 = oshw_htons(priMAC[1]);
+   	bp->sa2 = oshw_htons(priMAC[2]);
+   	bp->etype = oshw_htons(ETH_P_ECAT);
+}
+
+/** Get new frame identifier index and allocate corresponding rx buffer.
+ * @param[in] port        = port context struct
+ * @return new index.
+ */
+int ecx_getindex(ecx_portt *port)
+{
+   	int idx;
+   	int cnt = 0;
+
+	ee_port_lock();
+
+   	idx = port->lastidx + 1;
+   	/* index can't be larger than buffer array */
+   	if (idx >= EC_MAXBUF)
+      		idx = 0;
+
+   	/* try to find unused index */
+   	while ((port->rxbufstat[idx] != EC_BUF_EMPTY) && (cnt < EC_MAXBUF)) {
+      		idx++;
+      		cnt++;
+      		if (idx >= EC_MAXBUF)
+         		idx = 0;
+   	}
+   	port->rxbufstat[idx] = EC_BUF_ALLOC;
+   	if (port->redstate != ECT_RED_NONE)
+      		port->redport->rxbufstat[idx] = EC_BUF_ALLOC;
+   	port->lastidx = idx;
+
+	ee_port_unlock();
+
+   	return idx;
+}
+
+/** Set rx buffer status.
+ * @param[in] port        = port context struct
+ * @param[in] idx      = index in buffer array
+ * @param[in] bufstat  = status to set
+ */
+void ecx_setbufstat(ecx_portt *port, int idx, int bufstat)
+{
+   	port->rxbufstat[idx] = bufstat;
+   	if (port->redstate != ECT_RED_NONE)
+      		port->redport->rxbufstat[idx] = bufstat;
+}
+
+/** Transmit buffer over socket (non blocking).
+ * @param[in] port        = port context struct
+ * @param[in] idx         = index in tx buffer array
+ * @param[in] stacknumber  = 0=Primary 1=Secondary stack
+ * @return socket send result
+ */
+int ecx_outframe(ecx_portt *port, int idx, int stacknumber)
+{
+   	int lp;
+   	ec_stackT *stack;
+
+   	if (!stacknumber)
+      		stack = &(port->stack);
+   	else
+      		stack = &(port->redport->stack);
+   	lp = (*stack->txbuflength)[idx];
+   	(*stack->rxbufstat)[idx] = EC_BUF_TX;
+
+	eth_send_packet(port->dev_id, (*stack->txbuf)[idx], lp, 1);
+      	(*stack->rxbufstat)[idx] = EC_BUF_EMPTY;
```

Comment:
The buffer should be marked EC_BUF_EMPT if the send fail, otherwise it should stay EC_BUF_TX if some other thread is receiving it